src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,133 @@
+    private static final Object[] EMPTY_ELEMENT_DATA = {};
+    private static final Object[] DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA = {};
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA;
+    }
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            this.elementData = (T[]) new Object[initialCapacity];
+        } else if (initialCapacity == 0) {
+            this.elementData = (T[]) EMPTY_ELEMENT_DATA;
+        } else {
+            throw new IllegalArgumentException("Illegal Capacity: "
+                    + initialCapacity);
+        }
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private void rangeCheckForGetSetRemove(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private T[] grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA) {
+            int newCapacity = oldCapacity + (oldCapacity >> 1);
+            T[] newElementData = (T[]) new Object[newCapacity];
+            System.arraycopy(elementData, 0, newElementData, 0, size);
+            return newElementData;
+        } else {
+            return elementData = (T[]) new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+        }
+    }
+
+    private T[] grow() {
+        return grow(size);
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        if (size < elementData.length) {
+            elementData = grow(size + 1);
+        }
+        System.arraycopy(elementData, 0, elementData, size + 1, size + 1);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForGetSetRemove(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheckForGetSetRemove(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForGetSetRemove(index);
```

Comment:
```suggestion
        checkIndex(index);
```
Almost in all cases checkIndex is the same, move it a method, but in 1 case this check differs a little bit, just check it don't create method for it